src/main/java/dan200/computercraft/core/computer/ComputerThread.java

```diff
@@ -6,227 +6,304 @@
-                                catch( InterruptedException e )
-                                {
-                                }
-                            }
-                        }
-                    }
+                    s_thread.interrupt();
                 }
-            }, "Computer Dispatch Thread" );
-
-            m_thread.setDaemon(true);
-            m_thread.start();
-            m_running = true;
+            }
         }
     }
-    
-    public static void stop()
+
+    /**
+     * Queue a task to execute on the thread
+     *
+     * @param task     The task to execute
+     * @param computer The computer to execute it on, use {@code null} to execute on the default object.
+     */
+    public static void queueTask( ITask task, Computer computer )
     {
-        synchronized( m_lock )
+        Object queueObject = computer == null ? s_defaultOwner : computer;
+
+        BlockingQueue<ITask> queue;
+        synchronized( s_computerTaskQueues )
         {
-            if( m_running )
+            queue = s_computerTaskQueues.get( queueObject );
+            if( queue == null )
             {
-                m_stopped = true;
-                m_thread.interrupt();
+                s_computerTaskQueues.put( queueObject, queue = new LinkedBlockingQueue<ITask>( QUEUE_LIMIT ) );
+            }
+        }
+
+        synchronized( s_taskLock )
+        {
+            if( queue.offer( task ) && !s_computerTasksActiveSet.contains( queue ) )
+            {
+                s_computerTasksActive.add( queue );
+                s_computerTasksActiveSet.add( queue );
             }
         }
     }
-    
-    public static void queueTask( ITask _task, Computer computer )
+
+    /**
+     * Responsible for pulling and managing computer tasks. This pulls a task from {@link #s_computerTasksActive},
+     * creates a new thread using {@link TaskRunner} or reuses a previous one and uses that to execute the task.
+     *
+     * If the task times out, then it will attempt to interrupt the {@link TaskRunner} instance.
+     */
+    private static final class TaskExecutor implements Runnable
     {
-        Object queueObject = computer;
-        
-        if (queueObject == null)
+        private TaskRunner runner;
+        private Thread thread;
+
+        @Override
+        public void run()
         {
-            queueObject = m_defaultQueue;
+            try
+            {
+                while( true )
+                {
+                    // Wait for an active queue to execute
+                    BlockingQueue<ITask> queue = s_computerTasksActive.take();
+
+                    // If threads should be stopped then return
+                    synchronized( s_stateLock )
+                    {
+                        if( s_stopped ) return;
+                    }
+
+                    execute( queue );
+                }
+            }
+            catch( InterruptedException ignored )
+            {
+            }
```

Comment:
Is this really needed here? The thread is going to terminate at this point anyway it doesn't seem worth it. I might be wrong though.